### PR TITLE
🏃 Refactor all network client calls into networking package

### PR DESCRIPTION
**Pull request by macaptain**
This refactors all networking client behavior into the networking package. The network client has been removed from the compute Service struct, instead, we use the networking Service. Ports and trunking logic have been moved out of instance.go in the compute package into port.go and trunk.go in the networking package.

The process for refactoring is simple enough:

1. Delete network client from compute package
2. Move any functions using the network client to the networking package
3. Refactor the functions to call the networking client interface instead of gophercloud directly

**What this PR does / why we need it**:
This will allow us to write unit tests for ports and trunk functionality using the mocks in the networking package. It also considerably tidies up instance.go, since now all networking client calls are made in the networking package only.

**Which issue(s) this PR fixes**:
Continuing work on #843, finishes #883 after #935 was merged.

**Special notes for your reviewer**:
~~This is going to cause some rebasing to be required on open PRs around trunks and port security features, namely #934, #914, #921, but will at least allow these features to have unit tests written for them. I think the rebase should be straightforward, the `getOrCreatePort` function in instance.go has simply moved to `GetOrCreatePort` in the networking package. Alternatively, I don't mind waiting and rebasing this PR if those features are ready to go in.~~

PR is ready to go!

- [x] squashed commits

/hold

### CODE REVIEW

**Comment by macaptain** [comment]
Hi @hidekazuna, thanks for reviewing and merging #935. I saw you closed #883. Hopefully this PR does the same thing after #935.

/assign @hidekazuna

**Comment by jichenjc** [comment]
I am not sure how big trouble this can cause to other ongoing PR
maybe we need let those functional PRs merge then do the refactory all-in-one.. 
ok to any suggestions instead

**Comment by macaptain** [comment]
> I am not sure how big trouble this can cause to other ongoing PR
> maybe we need let those functional PRs merge then do the refactory all-in-one..
> ok to any suggestions instead

I agree with this, there is a lot of discussion on those other PRs and I don't want to complicate them any further. I am happy to hold the refactor until they are merged.

**Comment by hidekazuna** [comment]
> I agree with this, there is a lot of discussion on those other PRs and I don't want to complicate them any further. I am happy to hold the refactor until they are merged.

@macaptain Sorry late. Thank you for taking my PR into account. This PR looks fine for me as of now. It is OK to wait for other PRs not to bother them.

**Comment by macaptain** [comment]
Now I'm just waiting for #934 to be merged and then I'll rebase and this PR will be ready for review again.

**Comment by hidekazuna** [comment]
@macaptain Now that #934 has been merged, rebase please, if you do not mind.

File: pkg/cloud/services/networking/port.go
Comment (macaptain) [inline pkg/cloud/services/networking/port.go:57]: This is the only function I had to create (to avoid a circular dependency with `GetManagementPort`). All other functions in this file are moved from instance.go.

**Comment by macaptain** [comment]
> @macaptain Now that #934 has been merged, rebase please, if you do not mind.

@hidekazuna - I've rebased and this is ready for review again. Although the diff is quite large, the steps to make this change were not so complicated (they are in the PR description). I think the side-by-side diff will show what has moved. Thanks!

**Comment by hidekazuna** [comment]
> 
> 
> > @macaptain Now that #934 has been merged, rebase please, if you do not mind.
> 
> @hidekazuna - I've rebased and this is ready for review again. Although the diff is quite large, the steps to make this change were not so complicated (they are in the PR description). I think the side-by-side diff will show what has moved. Thanks!

@macaptain Thank s rebasing and advising to review. I will review in a few days.

**Comment by hidekazuna** [comment]
@jichenjc PTAL

**Comment by jichenjc** [comment]
I see some issue reported https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/993
which likely to be a regression of #935 and this PR seems to be a follow up of #935 

not sure whether this PR will make things worse or has conflict to the proposed fix of #993 
need more time to read the code and maybe @macaptain @mdbooth can help us in understand more

before that, I Think how about let's hold this?

**Comment by hidekazuna** [comment]
> before that, I Think how about let's hold this?

@jichenjc I agree to hold this PR. revisit this PR after #993 merged.

**Comment by mdbooth** [comment]
I haven't yet reviewed properly, but in general I'm very much in favour of continuing the refactor started by #935. The regression it introduced was caused by 2 very minor additions (checks for a zero length result) to a couple of networking functions which weren't strictly part of the refactor. However, it enables a pattern of unit testing which we can use to catch exactly this kind of bug in the future. I would like to see the same refactor in the compute package.

I suspect this change is going to conflict with *everything*, so I'd personally be in favour of merging it asap. I will try to review today.

File: pkg/cloud/services/networking/client.go
Comment (mdbooth) [inline pkg/cloud/services/networking/client.go:184]: This would be a regression because we're now logging an error in prometheus on delete of an object which has already been deleted where we previously were not. I think we need to make this, and all Delete* methods in this interface, call ObserveRequestsIgnoreNotFound.

We have the same issue in all Delete* methods.

File: pkg/cloud/services/networking/trunk.go
Comment (mdbooth) [inline pkg/cloud/services/networking/trunk.go:39]: Aside: I wonder how expensive this call is. We call it on every create and delete, but it doesn't seem like it's ever going to change. And if it did change, would it break inter-op with servers created before the change? What if it changed between create and delete? 🤔

Not for this PR!

File: pkg/cloud/services/networking/client.go
Comment (mdbooth) [inline pkg/cloud/services/networking/client.go:214]: Bikeshed: This should be plural: ListTrunks (along with all the other List* methods).

Leave it for now unless you want to change all the others in a separate commit, and it's obviously not super important! It does make me wince, though 😉

File: pkg/cloud/services/networking/trunk.go
Comment (mdbooth) [inline pkg/cloud/services/networking/trunk.go:83]: This is inconsistent. We should move this prometheus logging into the client, consistent with all the other logging.

That is going to inevitably change the logging from "trunk", "update" to something like "attributetags", "update" 🤔 I personally think that's ok, certainly less serious than the difference between logging errors and not logging errors on delete. Interested in other opinions, though.

File: pkg/cloud/services/networking/port.go
Comment (mdbooth) [inline pkg/cloud/services/networking/port.go:228]: nit: I personally wouldn't have moved this method. Its function is to iterate over servers.Network, which is a compute thing. Honestly I'd inline it in createInstance.

File: pkg/cloud/services/networking/port.go
Comment (mdbooth) [inline pkg/cloud/services/networking/port.go:239]: Not for this PR, but I'm pretty sure this is broken. This is listing ports which have the same name as the instance, but we don't create ports with the same name as the instance. I wonder if we should be deleting these by DeviceID instead 🤔

*NOT FOR THIS PR!*

File: pkg/cloud/services/compute/service.go
Comment (mdbooth) [inline pkg/cloud/services/compute/service.go:36]: 🥳

File: pkg/cloud/services/compute/instance.go
Comment (mdbooth) [inline pkg/cloud/services/compute/instance.go:199]: nit: As mentioned elsewhere, I would personally have left this method here.

**Review by mdbooth — CHANGES_REQUESTED** [review]
This mostly looks good to me. Just some minor issues which should be addressed before merging. Nits and bikesheds are up to you 😉

I would personally be in favour of merging this quickly despite the merge conflicts it will inevitably cause.

**Comment by jichenjc** [comment]
/approve
/hold cancel

ok, let's merge this given agreement and clearance of regression issue
and for all the nits, I guess we have chance to submit follow up PRs

**Review by macaptain — COMMENTED** [review]
Thanks for the thoughtful review @mdbooth ! I'll be addressing your points in follow-up PRs. I've raised #997 for making the metrics usage consistent.

@jichenjc - thanks for merging. This refactor means it's now possible to write unit tests for some of the network and ports logic, which will make this logic easier to maintain in future.

**Comment by macaptain** [comment]
@jichenjc oops, I managed to hide all my comments in response to @mdbooth :laughing:. I don't have permission to unhide them, so if you can unhide for me, that would be great.

**Comment by jichenjc** [comment]
@macaptain I don't know how to unhide them :( 
@hidekazuna do you know ?

**Pull request merged**
